lib/nice_partials/monkey_patch.rb

```diff
@@ -40,13 +46,44 @@ def p(*args)
   end
 
   def render(options = {}, locals = {}, &block)
-    _partial, @partial = partial, nice_partial
-    @partial.capture(block)
+    _partial = @partial
     super
   ensure
     @partial = _partial
   end
+
+  def _layout_for(*arguments, &block)
+    if block && !arguments.first.is_a?(Symbol)
+      partial.capture(*arguments, &block)
+    else
+      super
+    end
+  end
 end
 
-ActionView::Base.prepend NicePartials::RenderingWithLocalePrefix
-ActionView::Base.prepend NicePartials::RenderingWithAutoContext
+module NicePartials::PartialRendering
+  ActionView::PartialRenderer.prepend self
+
+  def render_partial_template(view, locals, template, layout, block)
+    view.partial.capture(&block) if block && !template.has_capturing_yield?
+    super
+  end
+end
+
+module NicePartials::CapturingYieldDetection
+  ActionView::Template.include self
+
+  # Matches yields that'll end up calling `capture`:
+  #   <%= yield %>
+  #   <%= yield something_else %>
+  #
+  # Doesn't match obfuscated `content_for` invocations, nor custom yields:
+  #   <%= yield :message %>
+  #   <%= something.yield %>
+  #
+  # Note: `<%= yield %>` becomes `yield :layout` with no `render` `block`, though this method assumes a block is passed.
+  def has_capturing_yield?
+    defined?(@has_capturing_yield) ? @has_capturing_yield :
+      @has_capturing_yield = source.match?(/\byield[\(? ]+(%>|[^:])/)
```

Comment:
Just noticing the `%>`: does this work with HAML?

Reply:
Originally I wanted to define this after template compilation, so we'd be Template Handler independent, but I just realized we'd need to trigger compilation for that to work (and that's why it didn't work).

I've never used anything besides ERB, how does HAML syntax go by the way? Like this?

```haml
%= yield
%= yield :title
```

Do you know what Slim looks like?

Reply:
I'm not a HAML/Slim user, but the `%>` got me thinking.

> I've never used anything besides ERB, how does HAML syntax go by the way? Like this?

Yep I think so.

Slim looks simlar http://slim-lang.com/